core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -582,35 +588,45 @@ public void run() {
     private RetryPlan makeRetryDecision(RetryPolicy retryPolicy, Status status, Metadata trailer) {
       boolean shouldRetry = false;
       long backoffInMillis = 0L;
-
-      if (retryPolicy.maxAttempts > substream.previousAttempts + 1) {
-        String pushbackStr = trailer.get(GRPC_RETRY_PUSHBACK_MS);
-        if (pushbackStr == null) {
-          if (retryPolicy.retryableStatusCodes.contains(status.getCode())) {
-            shouldRetry = true;
-            backoffInMillis = (long) (nextBackoffIntervalInSeconds * 1000D * random.nextDouble());
-            nextBackoffIntervalInSeconds = Math.min(
-                nextBackoffIntervalInSeconds * retryPolicy.backoffMultiplier,
-                retryPolicy.maxBackoffInSeconds);
-          } // else no retry
-        } else {
-          int pushback;
-          try {
-            pushback = Integer.parseInt(pushbackStr);
-          } catch (NumberFormatException e) {
-            pushback = -1;
-          }
-          if (pushback >= 0) {
-            shouldRetry = true;
-            backoffInMillis = pushback;
-            nextBackoffIntervalInSeconds = retryPolicy.initialBackoffInSeconds;
-          } // else no retry
+      String pushbackStr = trailer.get(GRPC_RETRY_PUSHBACK_MS);
+      if (pushbackStr == null) {
+        if (retryPolicy.retryableStatusCodes.contains(status.getCode())
+            && (throttle == null || throttle.onQualifiedFailureThenCheckIsAboveThreshold())
+            && retryPolicy.maxAttempts > substream.previousAttempts + 1) {
+          shouldRetry = true;
+          backoffInMillis = (long) (nextBackoffIntervalInSeconds * 1000D * random.nextDouble());
+          nextBackoffIntervalInSeconds = Math.min(
+              nextBackoffIntervalInSeconds * retryPolicy.backoffMultiplier,
+              retryPolicy.maxBackoffInSeconds);
+        } // else no retry
+      } else {
+        int pushback;
+        try {
+          pushback = Integer.parseInt(pushbackStr);
+        } catch (NumberFormatException e) {
+          pushback = -1;
         }
+        if (pushback < 0) { // no retry
+          if (throttle != null) {
+            throttle.onQualifiedFailureThenCheckIsAboveThreshold();
+          }
+        } else if (
+            (
+                retryPolicy.retryableStatusCodes.contains(status.getCode())
```

Comment:
The conditionals here are pretty complicated, especially with the side effects of invoking `#onQualifiedFailureThenCheckIsAboveThreshold()`.  It would be nice if the flow here could be made a little more straightforward. No simplification immediately jumps out, except FWIW you can factor out the `&& retryPolicy.maxAttempts > substream.previousAttempts + 1` condition since it's shared by both sides of the OR condition...that doesn't help a whole lot though. Turning that check into a boolean variable along with maybe `boolean retryableStatusCode = retryPolicy.retryableStatusCodes.contains(status.getCode())` might help somewhat, as that's also used in all of the conditionals here.

Reply:
Thanks for the suggestion! Done.